nntrainer/tensor/var_grad.cpp

```diff
@@ -16,19 +16,18 @@
 
 namespace nntrainer {
 
-Var_Grad::Var_Grad(const TensorDim &dim, bool train, const std::string &name) :
+Var_Grad::Var_Grad(const TensorDim &dim, bool train, bool alloc_now_, const std::string &name) :
```

Comment:
What's the difference from just having tensor dim and not allocated tensor?

Reply:
With alloc_now_ == true, when Var_Grad::initialize() is called, then the tensors memory is allocated.
With alloc_now_ == true, Var_Grad::initialize() will not allocated memory.

`NetworkGraph::inPlaceOptimization` takes place after Tensor in Var_Grad objects have been allocated and initialization has been done. But if the memory is already allocated with initialization, these later optimizations do not reduce the peak memory requirements.

Plus, by separating initialize() and allocate(), all the optimizations to decide how tensors across the layers interface (input and output connectivity between layers) is not required to done every time.

Reply:
Yes. I mean `Tensor::Tensor(Tensor&dim, nullptr, false)` vs `TensorDim` instead for non-allocated tensor. Because `allocation` happens always later, In this case `initialize` does not need to deal with `tensor(dim, null, false)` but just `tensordim` or `weights` that wraps around the `tensordim`.

Reply:
`Tensor::Tensor(Tensor&dim, nullptr, false)` vs `Tensor::Tensor(TensorDim)` are different as `alloc_now_` defaults to true.

Reply:
Yes, I mean between `Tensor::Tensor(Tensor&dim, nullptr, false)` vs `TensorDim`. And they could share weights (a wrapper of TensorDim in this stage) if needed.
Let's discuss this later as this would be slightly out of boundary of the PR

Reply:
I get what you mean. Semantically, they contain the same information. `Tensor::Tensor(Tensor&dim, nullptr, false)` is just a wrapper over `TensorDim` which will later contain memory.

Reply:
Yes and I think var_grad can do this. 
This could be discussed later when we need to reduce the complexity of tensor I guess.

Reply:
Yes, to reduce the overhead of tensor object.